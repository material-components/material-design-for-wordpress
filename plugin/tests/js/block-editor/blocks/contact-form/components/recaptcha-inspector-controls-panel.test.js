/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import '@testing-library/jest-dom/extend-expect';
import { render, fireEvent } from '@testing-library/react';
import { act } from 'react-dom/test-utils';

/**
 * Internal dependencies
 */
import RecaptchaInspectorControlsPanel from '../../../../../../assets/src/block-editor/blocks/contact-form/components/recaptcha-inspector-controls-panel';

/**
 * Render the component.
 *
 * @param {Object} props - Component props.
 * @return {Function} A functional component.
 */
const setup = props => {
	return render( <RecaptchaInspectorControlsPanel { ...props } /> );
};

const baseProps = {};

describe( 'RecaptchaInspectorControlsPanel', () => {
	let getData = 0;

	const ajax = jest.fn( args => {
		const response = {
			success: true,
		};
		const data = JSON.parse( args.data.data );

		if ( 'get' === data.action && getData < 2 ) {
			getData++;
			response.data = {
				material_design_recaptcha_client_secret: 'CLIENT_SECRET_FROM_AJAX',
				material_design_recaptcha_site_key: 'SITE_KEY_FROM_AJAX',
			};
		}

		return response;
	} );

	beforeAll( () => {
		global.materialDesign = {
			ajax_url: 'http://example.com/',
			recaptcha_ajax_nonce_action: 'CAPTCHA_NONCE',
		};
		global.jQuery = {
			ajax,
		};
	} );

	afterEach( () => {
		jest.restoreAllMocks();
		jest.clearAllMocks();
	} );

	it( 'matches snapshot', () => {
		const wrapper = setup( baseProps );
		expect( wrapper ).toMatchSnapshot();
	} );

	it( 'gets credentials by invoking ajax request', async () => {
		let inputs;

		await act( async () => {
			const props = { ...baseProps };
			const { container } = setup( props );
			inputs = container.querySelectorAll( '.components-text-control__input' );
		} );

		expect( ajax ).toHaveBeenCalledTimes( 1 );
		expect( ajax.mock.calls[ 0 ][ 0 ].data.data ).toContain( '"get"' );

		expect( inputs[ 0 ].value ).toStrictEqual( 'SITE_KEY_FROM_AJAX' );
		expect( inputs[ 1 ].value ).toStrictEqual( 'CLIENT_SECRET_FROM_AJAX' );
	} );

	it( "updates the site key and client secret in state when it's changed", () => {
		const setValue = jest.fn();
		const useStateSpy = jest.spyOn( React, 'useState' );
		useStateSpy.mockImplementation( initialValue => [
			initialValue,
			setValue,
		] );

		const props = { ...baseProps };
		const { container } = setup( props );
		const inputs = container.querySelectorAll(
			'.components-text-control__input'
		);

		act( () => {
			let mockEvent = {
				target: { value: 'SITE_KEY' },
			};
			fireEvent.change( inputs[ 0 ], mockEvent );

			mockEvent = {
				target: { value: 'CLIENT_SECRET' },
			};
			fireEvent.change( inputs[ 1 ], mockEvent );
		} );

		expect( setValue.mock.calls[ 0 ][ 0 ] ).toStrictEqual( 'SITE_KEY' );
		expect( setValue.mock.calls[ 1 ][ 0 ] ).toStrictEqual( 'CLIENT_SECRET' );
	} );

	it( 'saves credentials by invoking ajax request', async () => {
		const props = { ...baseProps };
		const { container } = setup( props );

		const inputs = container.querySelectorAll(
			'.components-text-control__input'
		);

		await act( async () => {
			let mockEvent = {
				target: { value: 'SITE_KEY' },
			};
			await fireEvent.change( inputs[ 0 ], mockEvent );

			mockEvent = {
				target: { value: 'CLIENT_SECRET' },
			};
			await fireEvent.change( inputs[ 1 ], mockEvent );

			const saveBtn = container.querySelector( '.save-button' );
			await fireEvent.click( saveBtn );
		} );

		expect( ajax ).toHaveBeenCalledTimes( 2 );
		expect( ajax.mock.calls[ 0 ][ 0 ].data.data ).toContain( '"get"' );
		expect( ajax.mock.calls[ 1 ][ 0 ].data.data ).toContain( '"save"' );
		expect( ajax.mock.calls[ 1 ][ 0 ].data.data ).toContain( '"SITE_KEY"' );
		expect( ajax.mock.calls[ 1 ][ 0 ].data.data ).toContain(
			'"CLIENT_SECRET"'
		);
	} );

	it( 'does not save credentials if key and secret are not provided', async () => {
		const props = { ...baseProps };
		const { container } = setup( props );

		const inputs = container.querySelectorAll(
			'.components-text-control__input'
		);

		await act( async () => {
			let mockEvent = {
				target: { value: '' },
			};
			await fireEvent.change( inputs[ 0 ], mockEvent );

			mockEvent = {
				target: { value: '' },
			};
			await fireEvent.change( inputs[ 1 ], mockEvent );

			const saveBtn = container.querySelector( '.save-button' );
			await fireEvent.click( saveBtn );
		} );

		expect( ajax ).toHaveBeenCalledTimes( 1 );
	} );

	it( 'clears credentials by invoking ajax request', async () => {
		const props = { ...baseProps };
		const { container } = setup( props );

		const inputs = container.querySelectorAll(
			'.components-text-control__input'
		);

		await act( async () => {
			let mockEvent = {
				target: { value: 'SITE_KEY' },
			};
			await fireEvent.change( inputs[ 0 ], mockEvent );

			mockEvent = {
				target: { value: 'CLIENT_SECRET' },
			};
			await fireEvent.change( inputs[ 1 ], mockEvent );

			const saveBtn = container.querySelector( '.clear-button' );
			await fireEvent.click( saveBtn );
		} );

		expect( ajax ).toHaveBeenCalledTimes( 2 );
		expect( ajax.mock.calls[ 0 ][ 0 ].data.data ).toContain( '"get"' );
		expect( ajax.mock.calls[ 1 ][ 0 ].data.data ).toContain( '"clear"' );
	} );
} );
