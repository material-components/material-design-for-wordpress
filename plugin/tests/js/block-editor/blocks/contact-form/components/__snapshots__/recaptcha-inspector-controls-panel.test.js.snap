// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecaptchaInspectorControlsPanel matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <p
      class="a11y-speak-intro-text"
      hidden="hidden"
      id="a11y-speak-intro-text"
      style="position: absolute;margin: -1px;padding: 0;height: 1px;width: 1px;overflow: hidden;clip: rect(1px, 1px, 1px, 1px);-webkit-clip-path: inset(50%);clip-path: inset(50%);border: 0;word-wrap: normal !important;"
    >
      Notifications
    </p>
    <div
      aria-atomic="true"
      aria-live="assertive"
      aria-relevant="additions text"
      class="a11y-speak-region"
      id="a11y-speak-assertive"
      style="position: absolute;margin: -1px;padding: 0;height: 1px;width: 1px;overflow: hidden;clip: rect(1px, 1px, 1px, 1px);-webkit-clip-path: inset(50%);clip-path: inset(50%);border: 0;word-wrap: normal !important;"
    />
    <div
      aria-atomic="true"
      aria-live="polite"
      aria-relevant="additions text"
      class="a11y-speak-region"
      id="a11y-speak-polite"
      style="position: absolute;margin: -1px;padding: 0;height: 1px;width: 1px;overflow: hidden;clip: rect(1px, 1px, 1px, 1px);-webkit-clip-path: inset(50%);clip-path: inset(50%);border: 0;word-wrap: normal !important;"
    />
    <div>
      <div
        class="components-panel__body is-opened"
      >
        <h2
          class="components-panel__body-title"
        >
          <button
            aria-expanded="true"
            class="components-button components-panel__body-toggle"
            type="button"
          >
            <span
              aria-hidden="true"
            >
              <svg
                aria-hidden="true"
                class="components-panel__arrow"
                focusable="false"
                height="24"
                role="img"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 8l-6 5.4 1 1.2 5-4.6 5 4.6 1-1.2z"
                />
              </svg>
            </span>
            Google reCAPTCHA v3
          </button>
        </h2>
        <p>
          Add your reCAPTCHA site and secret keys to protect your form from spam.
        </p>
        <a
          class="components-external-link"
          href="https://g.co/recaptcha/v3"
          rel="external noreferrer noopener"
          target="_blank"
        >
          Generate keys
          <span
            class="components-visually-hidden"
          >
            (opens in a new tab)
          </span>
          <svg
            aria-hidden="true"
            class="components-external-link__icon"
            focusable="false"
            height="24"
            role="img"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M18.2 17c0 .7-.6 1.2-1.2 1.2H7c-.7 0-1.2-.6-1.2-1.2V7c0-.7.6-1.2 1.2-1.2h3.2V4.2H7C5.5 4.2 4.2 5.5 4.2 7v10c0 1.5 1.2 2.8 2.8 2.8h10c1.5 0 2.8-1.2 2.8-2.8v-3.6h-1.5V17zM14.9 3v1.5h3.7l-6.4 6.4 1.1 1.1 6.4-6.4v3.7h1.5V3h-6.3z"
            />
          </svg>
        </a>
        <a
          class="components-external-link"
          href="https://developers.google.com/recaptcha/docs/v3"
          rel="external noreferrer noopener"
          target="_blank"
        >
          Documentation
          <span
            class="components-visually-hidden"
          >
            (opens in a new tab)
          </span>
          <svg
            aria-hidden="true"
            class="components-external-link__icon"
            focusable="false"
            height="24"
            role="img"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M18.2 17c0 .7-.6 1.2-1.2 1.2H7c-.7 0-1.2-.6-1.2-1.2V7c0-.7.6-1.2 1.2-1.2h3.2V4.2H7C5.5 4.2 4.2 5.5 4.2 7v10c0 1.5 1.2 2.8 2.8 2.8h10c1.5 0 2.8-1.2 2.8-2.8v-3.6h-1.5V17zM14.9 3v1.5h3.7l-6.4 6.4 1.1 1.1 6.4-6.4v3.7h1.5V3h-6.3z"
            />
          </svg>
        </a>
        <br />
        <br />
        <div
          class="components-base-control"
        >
          <div
            class="components-base-control__field"
          >
            <label
              class="components-base-control__label"
              for="inspector-text-control-0"
            >
              Site Key
            </label>
            <input
              class="components-text-control__input"
              id="inspector-text-control-0"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="components-base-control"
        >
          <div
            class="components-base-control__field"
          >
            <label
              class="components-base-control__label"
              for="inspector-text-control-1"
            >
              Client Secret
            </label>
            <input
              class="components-text-control__input"
              id="inspector-text-control-1"
              type="text"
              value=""
            />
          </div>
        </div>
        <p>
          <i
            class="material-icons tab-add__icon"
          >
            warning
          </i>
          Google reCAPTCHA will only work and be rendered on the frontend
        </p>
        <br />
        <button
          class="components-button save-button is-primary"
          style="margin-right: 20px;"
          type="button"
        >
          Save
        </button>
        <button
          class="components-button clear-button is-primary"
          type="button"
        >
          Clear
        </button>
        <br />
        <br />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="components-panel__body is-opened"
    >
      <h2
        class="components-panel__body-title"
      >
        <button
          aria-expanded="true"
          class="components-button components-panel__body-toggle"
          type="button"
        >
          <span
            aria-hidden="true"
          >
            <svg
              aria-hidden="true"
              class="components-panel__arrow"
              focusable="false"
              height="24"
              role="img"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 8l-6 5.4 1 1.2 5-4.6 5 4.6 1-1.2z"
              />
            </svg>
          </span>
          Google reCAPTCHA v3
        </button>
      </h2>
      <p>
        Add your reCAPTCHA site and secret keys to protect your form from spam.
      </p>
      <a
        class="components-external-link"
        href="https://g.co/recaptcha/v3"
        rel="external noreferrer noopener"
        target="_blank"
      >
        Generate keys
        <span
          class="components-visually-hidden"
        >
          (opens in a new tab)
        </span>
        <svg
          aria-hidden="true"
          class="components-external-link__icon"
          focusable="false"
          height="24"
          role="img"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18.2 17c0 .7-.6 1.2-1.2 1.2H7c-.7 0-1.2-.6-1.2-1.2V7c0-.7.6-1.2 1.2-1.2h3.2V4.2H7C5.5 4.2 4.2 5.5 4.2 7v10c0 1.5 1.2 2.8 2.8 2.8h10c1.5 0 2.8-1.2 2.8-2.8v-3.6h-1.5V17zM14.9 3v1.5h3.7l-6.4 6.4 1.1 1.1 6.4-6.4v3.7h1.5V3h-6.3z"
          />
        </svg>
      </a>
      <a
        class="components-external-link"
        href="https://developers.google.com/recaptcha/docs/v3"
        rel="external noreferrer noopener"
        target="_blank"
      >
        Documentation
        <span
          class="components-visually-hidden"
        >
          (opens in a new tab)
        </span>
        <svg
          aria-hidden="true"
          class="components-external-link__icon"
          focusable="false"
          height="24"
          role="img"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18.2 17c0 .7-.6 1.2-1.2 1.2H7c-.7 0-1.2-.6-1.2-1.2V7c0-.7.6-1.2 1.2-1.2h3.2V4.2H7C5.5 4.2 4.2 5.5 4.2 7v10c0 1.5 1.2 2.8 2.8 2.8h10c1.5 0 2.8-1.2 2.8-2.8v-3.6h-1.5V17zM14.9 3v1.5h3.7l-6.4 6.4 1.1 1.1 6.4-6.4v3.7h1.5V3h-6.3z"
          />
        </svg>
      </a>
      <br />
      <br />
      <div
        class="components-base-control"
      >
        <div
          class="components-base-control__field"
        >
          <label
            class="components-base-control__label"
            for="inspector-text-control-0"
          >
            Site Key
          </label>
          <input
            class="components-text-control__input"
            id="inspector-text-control-0"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="components-base-control"
      >
        <div
          class="components-base-control__field"
        >
          <label
            class="components-base-control__label"
            for="inspector-text-control-1"
          >
            Client Secret
          </label>
          <input
            class="components-text-control__input"
            id="inspector-text-control-1"
            type="text"
            value=""
          />
        </div>
      </div>
      <p>
        <i
          class="material-icons tab-add__icon"
        >
          warning
        </i>
        Google reCAPTCHA will only work and be rendered on the frontend
      </p>
      <br />
      <button
        class="components-button save-button is-primary"
        style="margin-right: 20px;"
        type="button"
      >
        Save
      </button>
      <button
        class="components-button clear-button is-primary"
        type="button"
      >
        Clear
      </button>
      <br />
      <br />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
